pkg/operator/noobaa/system_state.go

```diff
@@ -0,0 +1,698 @@
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package noobaa
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"text/template"
+
+	dockerref "github.com/docker/distribution/reference"
+	semver "github.com/hashicorp/go-version"
+	nb_v1 "github.com/rook/rook/pkg/apis/noobaa.rook.io/v1alpha1"
+	"github.com/rook/rook/pkg/operator/noobaa/nbapi"
+	apps_v1 "k8s.io/api/apps/v1"
+	core_v1 "k8s.io/api/core/v1"
+	rbac_v1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// SystemState keeps the desired/current state of a noobaa system and its resources
+type SystemState struct {
+	Operator        *Operator
+	CachedSystem    *nb_v1.NooBaaSystem
+	System          *nb_v1.NooBaaSystem
+	ServiceAccount  *core_v1.ServiceAccount
+	Role            *rbac_v1.Role
+	RoleBinding     *rbac_v1.RoleBinding
+	CoreApp         *apps_v1.StatefulSet
+	ServiceMgmt     *core_v1.Service
+	ServiceS3       *core_v1.Service
+	SecretOp        *core_v1.Secret
+	SecretAdmin     *core_v1.Secret
+	NBClient        *nbapi.RPCClient
+	ErrorSuppressed bool
+}
+
+func NewSystemState(operator *Operator, cachedSys *nb_v1.NooBaaSystem) *SystemState {
+
+	sys := cachedSys.DeepCopy()
+	serviceAccountName := sys.Name + "-service-account"
+	roleName := sys.Name + "-role"
+	roleBindingName := sys.Name + "-role-binding"
+	coreAppName := sys.Name + "-core"
+	ServiceMgmtName := sys.Name + "-mgmt"
+	ServiceS3Name := "s3" // TODO: handle collision in namespace
+	operatorSecretName := sys.Name + "-operator"
+	adminSecretName := sys.Name + "-admin"
+	logdirPVCName := sys.Name + "-logdir"
+	datadirPVCName := sys.Name + "-datadir"
+	corePodSelectorLabels := map[string]string{
+		"noobaa-core": coreAppName,
+	}
+	mgmtSelectorLabels := map[string]string{
+		"noobaa-mgmt": coreAppName,
+	}
+	s3SelectorLabels := map[string]string{
+		"noobaa-s3": coreAppName,
+	}
+	corePodLabels := map[string]string{
+		"noobaa-core": coreAppName,
+		"noobaa-s3":   coreAppName,
+		"noobaa-mgmt": coreAppName,
+	}
+	prometheusAnnotations := map[string]string{
+		"prometheus.io/scrape": "true",
+		"prometheus.io/scheme": "http",
+		"prometheus.io/port":   "8080",
+	}
+
+	if sys.Spec.Image == "" {
+		sys.Spec.Image = ContainerImage
+	}
+
+	makeInt32Ptr := func(x int32) *int32 { return &x }
+
+	makeObjectMeta := func(name string) meta_v1.ObjectMeta {
+		ownerIsController := true
+		owner := meta_v1.OwnerReference{
+			APIVersion: sys.APIVersion,
```

Comment:
There was a bug in the owner references creation that you'll want to pick up on this line with `APIVersion`. See #3350.

Reply:
I am copying the APIVersion from the NooBaaSystem object, so I think it's good. Thanks.